python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -688,8 +703,191 @@ def asset_check_spec(self) -> AssetCheckSpec:
         return asset_checks_def.spec
 
 
-# actually forking the object type for assets is tricky for users in the cases of:
-#  * manually constructing ops to make AssetsDefinitions
-#  * having ops in a graph that form a graph backed asset
-# so we have a single type that users can call by their preferred name where appropriate
-AssetExecutionContext: TypeAlias = OpExecutionContext
+OP_EXECUTION_CONTEXT_ONLY_METHODS = set(
+    [
+        "describe_op",
+        "file_manager",
+        "has_assets_def",
+        "get_mapping_key",
+        # "get_step_execution_context", # used by internals
+        "job_def",
+        "job_name",
+        "node_handle",
+        "op",
+        "op_config",
+        # "op_def", # used by internals
+        "op_handle",
+        "retry_number",
+        "step_launcher",
+        # "has_events", # used by internals
+        "consumer_events",
+    ]
+)
+
+
+PARTITION_KEY_RANGE_AS_ALT = "use partition_key_range or partition_key_range_for_asset instead"
+INPUT_OUTPUT_ALT = "not use input or output names and instead use asset keys directly"
+OUTPUT_METADATA_ALT = "return MaterializationResult from the asset instead"
+
+DEPRECATED_IO_MANAGER_CENTRIC_CONTEXT_METHODS = {
+    "add_output_metadata": OUTPUT_METADATA_ALT,
+    "asset_key_for_input": INPUT_OUTPUT_ALT,
+    "asset_key_for_output": INPUT_OUTPUT_ALT,
+    "asset_partition_key_for_input": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partition_key_for_output": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partition_key_range_for_input": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partition_key_range_for_output": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partition_keys_for_input": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partition_keys_for_output": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partitions_time_window_for_input": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partitions_time_window_for_output": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partitions_def_for_input": PARTITION_KEY_RANGE_AS_ALT,
+    "asset_partitions_def_for_output": PARTITION_KEY_RANGE_AS_ALT,
+    "get_output_metadata": "use op_execution_context.op_def.get_output(...).metadata",
+    "merge_output_metadata": OUTPUT_METADATA_ALT,
+    "output_for_asset_key": INPUT_OUTPUT_ALT,
+    "selected_output_names": INPUT_OUTPUT_ALT,
+}
+
+ALTERNATE_AVAILABLE_METHODS = {
+    "has_tag": "use dagster_run.has_tag instead",
+    "get_tag": "use dagster_run.get_tag instead",
+    "run_tags": "use dagster_run.tags instead",
+    "set_data_version": "use MaterializationResult instead",
+}
+
+
+class AssetExecutionContext:
```

Comment:
Thread for figuring out why `AssetExecutionContext` was not made a subclass of `OpExecutionContext` to begin with. 

Background:
* `AssetExecutionContext` began as a type alias to align on naming and get a quick docs/examples improvement ([here](https://github.com/dagster-io/dagster/pull/14656))
* `AssetExecutionContext` was made a subclass of `OpExecutionContext` [here](https://github.com/dagster-io/dagster/pull/14814)
* `AssetExecutionContext` was reverted back to a type alias [here](https://github.com/dagster-io/dagster/pull/15032)

In the revert PR, the reasoning for reverting was:
```
Conditions like:

* manually constructing AssetsDefinition with a manually written @op
* @ops that make up a graph backed AssetsDefinition

make having different context objects trickier for users than originally anticipated.
```

There is also a [slack thread](https://elementl-workspace.slack.com/archives/C03CA4TVCAW/p1692378948383339?thread_ts=1692309846.639679&cid=C03CA4TVCAW) mentioning this where alex says 
```
the wall I hit with trying to split AssetExecutionContext and OpExecutionContext was resolving what the ops in a graph-backed-asset should receive.
```

based on this, my interpretation is that the issue wasn't a technical one (python limitation, inability to pass the correct context through, etc), but more a design issue "What is the correct context for an `@op` in a `@graph_backed_asset` to receive?"

Reply:
Ok got it. Thanks for digging that up.